Add Class B Support

snelson-senet opened the pull request:
### Description
Brief Class B Summary
•	Core functionality encapsulated in class LoRaMacClassB
o	Beacon acquisition & tracking 
o	Unicast and Multicast Ping slot scheduling
o	Conditional compilation controlled by mbed_lib.json option
•	All regions supported
o	Tested regions: US915, EU868, IN865
•	Added MAC Commands
o	PingSlotInfo
o	 PingSlotChannelReq
o	BeaconFreqReq
 
Existing Mbed LoRaWAN Stack Design Changelog
1.	LoRaWAN GPS time resolution changed from seconds to milliseconds.  Millisecond precision is required for beacon and ping slot scheduling
2.	Added beacon event processing to the stack layer
3.	Added Class B radio  callbacks
4.	Added Class B pause/resume around class A uplink transmission and receive windows
5.	Changed DeviceTimeReq to sticky MAC command
o	I implemented this command as sticky before integration with Hasnain's version. I like the sticky behavior and it "stuck", for now. I'll let those with greater voting rights than me to decide the outcome. My 2 cents is that the common scenario is the application will want the stack to continue sending the command until the response is received.
6.	Added PHY layer beacon and ping slot window configuration
7.	While working on the PHY layer Class B support,  I found that where there were region derived LoRaPHY::rx_config()  implementations, they could be easily removed with minor modification to the base implementation, thus simplifying the design by removing duplicated code. 
8.	Added time on air computation  to LoRaRadio class for beacon time on air calculation. 
o	Current time on air routines implemented by each sx12xx driver do not support calculating received beacon time on air because the computation inputs are the current radio transmit configuration.


### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@hasnainvirk 


### Release Notes

Commit:
Adding network assisted GPS time acquisition

LoRaWANTimer class is responsible for providing time base for the
LoRaWAN stack in Mbed OS. It derives its core monotonic clock from the
EventQueue which is essentially a millisecond ticker or equivalent (in
tickless mode). LoRaWAN network can assist the device for device level
time synchronization using ClockSynchronization protocol (for v1.0.2) or
by replying to DevTimeReq MAC command (for v1.0.3 and above). The time
base used for device synchronization using above methods is purely
based on GPS epoch and needs no conversion to UTC or TAI etc (no
adjustment for leap seconds either).
Therefore we have extended the LoRaWANTimer APIs to account for GPS time
base if assisted by the network.
get_gps_time() is the API to use to acquire GPS time base from the
stack. If the network never assisted the device with a time stamp, a
value of zero is returned which means GPS time base is not available.
set_gps_time(time) API is used to set a given GPS time-stamp when the
network sends it.
These APIs are hooked via LoRaMAC to LoRaWANInterface which exposes the
access to the application.

Commit:
Exposing GPS/UTC time APIs

Provides hooks for the internal APIs handling GPS time base for device
synchronization.
In addition to that 'set_system_time_utc(...)' API is added which have no
roots in the stack. Its just a helper for the application developer to
acquire UTC time base given the difference between TAI and UTC time
bases.

Commit:
Fixing DevTimeAns handling

As per discussion with Nicolas Sornin, any device time synchronization
value must be handled as is (monotonically increasing GPS epoch time guaranteed to be monotonic with granularity 1 second).

Once we receive a response to DevTimeReq (i.e., DevTimeAns is received)
we immediately set our local gps time stamp and take a snapshot of our
local monotonic (ticker or equivalent) clock. This is achieved by
calling 'set_gps_time(...)' API in the LoRaWANTimer module.
After that an event is generated to let the application know that a
device time has been synched.

snelson-senet committed:
Switch to millisecond resolution  gps time

The existing gps time  resolution is seconds, while class b beacon and ping slot scheduling requires millisecond resolution.
This set of changes switches gps time to  millisecond resolution.

snelson-senet committed:
Add Crypto ping slot random offset computation 

Implement ping slot pseudo-random offset computation

snelson-senet committed:
Add Class B related MAC Commands

-  PingSlotInfoReq & PingSlotInfoAns 
-  PingSlotChannelReq & PingSlotChannelAns
-  BeaconFreqReq & BeaconFreqAns

Switch to millisecond computation for  device time handling

snelson-senet committed:
LoRaWAN Interface Class B APIs

Added:
-  enable beacon acquisition
-  add/remove PingSlotInfoReq MAC  command
-  get last received beacon

snelson-senet committed:
Class B json configuration

Added Class B configuration

snelson-senet committed:
PHY Class B beacon and ping slot configuration

- Add PHY beacon and ping slot configuration

snelson-senet committed:
PHY Class B handling

Support class b beacon and ping slot rx configuration

snelson-senet committed:
PHY Class B configuration

- Initialize Class B configuration
- Implement class B receiver configuration
- For regions such as US915 with derived rx_config(),  changes to the base class implementation makes the derivation unnecessary , so these region specific versions have been removed

snelson-senet committed:
Add core Class B  implementation 

Here it is - the main Class B functionaiity

snelson-senet committed:
LoRaRadio on air time computation

Instead of implementing this in each derived radio class, let's do it once here.

snelson-senet committed:
Stack Class B support

- Switched to millisecond gps time
- Updated process rx/rx_timeout events for  class B windows
- Added beacon event callback

snelson-senet committed:
Fix class b disable timers

Stop beacon acquisition timer

snelson-senet committed:
Reorder class b slot types by priority

snelson-senet committed:
Fix PingSlotInfoReq parsing

PingSlotInfoReq 1 byte payload

snelson-senet committed:
Moved Beacon GPS time set

Moved from stack to class b beacon handler

snelson-senet committed:
Fix LoRaWAN interface gps time type

Switched to  high precision gps time

snelson-senet committed:
Remove extra beacon debug setting

Was only used to print bad beacon frames

snelson-senet committed:
Add back PHY computed ping slot window offset 

Removed unintentionally in last commit

snelson-senet committed:
mbed-os style update: Add LoRaMacClassBInterface 

Make class function names lower case. Previously, class b's public  interface was part of the class and to avoid function name conflicts the static API was being distinguished from instance methods  by upper-casing the first letter. 
Fixed by moving the public interface to new LoRaMacClassBInterface

snelson-senet committed:
Add back Class B Rx handler skip optimization

Skip handler check was dropped on the last commit moving the static class b api out of its class to an interface.

snelson-senet committed:
Astyle format ting

99.9% white space changes

snelson-senet committed:
Doxygen formatting

Reviewed and fixed  potential doxygen format issues.

snelson-senet committed:
Radio not put to sleep after class A Rx windows.

The Rx done/timeout event handler now checks if the radio is currently
transmitting before putting the radio to sleep. This test returns true in
class A rx window event handlers, so the radio is not not put to sleep.

Now to accurately test if transmitting, a mod ongoing flag is set true on send
and false on tx done/timeout.

0xc0170 commented on features/lorawan/lorastack/phy/LoRaPHY.cpp:
```diff
@@ -886,30 +894,74 @@ void LoRaPHY::compute_rx_win_params(int8_t datarate, uint8_t min_rx_symbols,
+        case RX_SLOT_WIN_CLASS_C:
+            break;
+        default:
+            MBED_ASSERT(false);
```
shouldnt these (also similar on line 964) use `MBED_ERROR` ? to report what error and msg as well ?

snelson-senet replied:
After reading this again, I  see I went a different direction by adding a return status that requires the caller to  report an error. I'm unfamiliar with MBED_ERROR, so I'll look into that.

0xc0170 commented on features/lorawan/lorastack/mac/LoRaMacClassBInterface.cpp:
```diff
@@ -0,0 +1,5 @@
+#include "LoRaMacClassBInterface.h"
```
this file needs license header file

snelson-senet replied:
Committed

0xc0170 commented on features/lorawan/lorastack/mac/LoRaMacClassBInterface.h:
```diff
@@ -0,0 +1,167 @@
+#ifndef MBED_LORAWAN_LORAMACCLASSB_INTERFACE_H_
+#define MBED_LORAWAN_LORAMACCLASSB_INTERFACE_H_
```
please move these 2 lines after the license header file

snelson-senet replied:
Done!

0xc0170 commented on features/lorawan/lorastack/mac/LoRaMacClassBInterface.h:
```diff
@@ -0,0 +1,167 @@
+#ifndef MBED_LORAWAN_LORAMACCLASSB_INTERFACE_H_
+#define MBED_LORAWAN_LORAMACCLASSB_INTERFACE_H_
+/**
+ * Copyright (c) 2017, Arm Limited and affiliates.
```
year 2019 for new files

snelson-senet replied:
All set

snelson-senet committed:
License headers

- 2019 for new file
- Added license where missing

snelson-senet committed:
Add a return status to Rx window computation functions

- Added a bool return status to  LoRaPHY::compute_*_win_params(...)   methods. Currently, failed status returned if the rx slot type is invalid or rf frequency is invalid
 - Updated Class B implementation to check return status

snelson-senet committed:
Move class B resume and close rx window improvement 

Use close_rx_window  callback to disable active class b window. This change protects against disabling the radio during class A transmit 
Move class B resume notification to one location in LoRaMac::set_tx_ongoing()

0xc0170 commented on features/lorawan/lorastack/phy/LoRaPHY.h:
```diff
@@ -591,6 +628,70 @@ class LoRaPHY : private mbed::NonCopyable<LoRaPHY> {
+
+    /**
+     * @brief compute_beacon_time_on_air
+     * @param compute_beacon_time_on_air computes beacon frame time on air
```
param - there's no parameter, brief should contain this description?

snelson-senet replied:
Fixed

0xc0170 commented on features/lorawan/lorastack/phy/LoRaPHY.h:
```diff
@@ -591,6 +628,70 @@ class LoRaPHY : private mbed::NonCopyable<LoRaPHY> {
+     *                              in a [-rxError : +rxError] ms interval around
+     *                              RxOffset.
+     *
+     * @param [out] rx_conf_params  Pointer to the structure that needs to be
```
this is config ?

snelson-senet replied:
Fixed

0xc0170 commented on features/lorawan/lorastack/phy/LoRaPHY.h:
```diff
@@ -591,6 +628,70 @@ class LoRaPHY : private mbed::NonCopyable<LoRaPHY> {
+    /**
+     * @brief get_beacon_rfu_size Gets the current region beacon RFU field sizes
+     */
+    virtual void get_beacon_rfu_size(uint8_t &rfu1, uint8_t &rfu2);
```
why these param are not documented ? similar to what is on line 665 (param in/out) ?

snelson-senet replied:
Fixed

0xc0170 commented on features/lorawan/lorastack/phy/LoRaPHY.h:
```diff
@@ -591,6 +628,70 @@ class LoRaPHY : private mbed::NonCopyable<LoRaPHY> {
+     *                              in a [-rxError : +rxError] ms interval around
+     *                              RxOffset.
+     *
+     * @param [out] rx_conf_params  Pointer to the structure that needs to be
```
config here

snelson-senet replied:
Fixed

0xc0170 commented on features/lorawan/lorastack/phy/LoRaPHY.h:
```diff
@@ -411,6 +412,42 @@ class LoRaPHY : private mbed::NonCopyable<LoRaPHY> {
+     * @brief accept_ping_slot_channel_req  Makes decision whether to accept
+     *        or reject PIngSlotChannelReq MAC command.
+     *
+     * @param frequency The unicast ping slot channel
```
datarate not documented?

snelson-senet replied:
Added

0xc0170 commented on features/lorawan/lorastack/mac/LoRaMacCommand.h:
```diff
@@ -161,14 +162,23 @@ class LoRaMacCommand {
+     *                                  LORAWAN_STATUS_LENGTH_ERROR: Buffer full,
+     *                                  LORAWAN_STATUS_PARAMETER_INVALID: Invalid periodicity
+     */
+    lorawan_status_t add_ping_slot_info_req(uint8_t periodicity);
```
param not needed to be documented?

snelson-senet replied:
Updated

0xc0170 reviewed (changes requested):
Few documentation questions

snelson-senet committed:
Function documentation cleanup

snelson-senet committed:
Fix Travis astyle failures

Ran asytle to fix line ending s

snelson-senet committed:
astyle formatting

snelson-senet committed:
Another astyle format change

snelson-senet committed:
Add LoRaWANInterface Class B API related documentation

- Documented LoRaWANInterface beacon acquisition,  and ping slot APIs
- Existing GPS time set/get documentation was only in LoRaWANStack.h, so I copied their docs to LoRaWANInterface and lightly modified them 
  to be a higher level description like the other documentation in this file.
- Updated  set_utc_time to handle millisecond precision gps time.

snelson-senet committed:
Fix unit tests 

Fix compilation failures caused by lorawan api changes introduced by Class B support.

snelson-senet committed:
Unit Test LoRaWANStack:  Remove set class B device class test  

With class B now being available the device class test moved to the LoRaMac class which then queries the Class B interface for support.  
The Class B interface is not available here because the unit test stack class instantiates a LoRaMac stub class .

snelson-senet committed:
Unit Test LoRaPHY:  rx_config test updates

LoRaPHY::rx_config() is now used by all regions. All region specific implementations have been removed.

snelson-senet committed:
Unit Test LoRaPHY

Fix compilation failures caused by lorawan api changes introduced by Class B support

snelson-senet committed:
Astyle fixes

astyle on unit test sources

snelson-senet committed:
Fix IN865 Beacon Frequency 

Set IN865 beacon frequency to 866.55 MHz

snelson-senet committed:
Fix Beacon Time On Air Computation

Beacon time on air  computation bandwidth param was hard coded to 500 KHz.  Now  it is read from the phy bandwidths table.

snelson-senet committed:
Fix  get_rx1_frequency

Return alternate RX1 if set

snelson-senet committed:
Unit test fix

Add missing Class B flags to unit test cmake

AnttiKauppila commented on UNITTESTS/stubs/LoRaMac_stub.cpp:
```diff
@@ -506,3 +507,34 @@ lorawan_status_t LoRaMac::rejoin(join_req_type_t rejoin_type, bool is_forced, ui
+
+void LoRaMac::set_gps_time(lorawan_gps_time_t gps_time)
+{
+    _lora_time.set_gps_time(gps_time);
```
This line is not needed

AnttiKauppila commented on UNITTESTS/stubs/LoRaMac_stub.cpp:
```diff
@@ -506,3 +507,34 @@ lorawan_status_t LoRaMac::rejoin(join_req_type_t rejoin_type, bool is_forced, ui
+
+lorawan_gps_time_t LoRaMac::get_gps_time(void)
+{
+    return _lora_time.get_gps_time();
```
You should return stubbed value here

AnttiKauppila commented on UNITTESTS/stubs/LoRaWANStack_stub.cpp:
```diff
@@ -305,3 +305,36 @@ void LoRaWANStack::process_idle_state(lorawan_status_t &op_status)
+
+void LoRaWANStack::set_current_gps_time(lorawan_gps_time_t gps_time)
+{
+    _loramac.set_gps_time(gps_time);
```
Remove this line

AnttiKauppila commented on UNITTESTS/stubs/LoRaWANStack_stub.cpp:
```diff
@@ -305,3 +305,36 @@ void LoRaWANStack::process_idle_state(lorawan_status_t &op_status)
+
+lorawan_gps_time_t LoRaWANStack::get_current_gps_time()
+{
+    return _loramac.get_gps_time();
```
return stubbed value

AnttiKauppila reviewed (changes requested):
@snelson-senet Thanks for your changes. Please fix the stubs to not have implementation in them.

snelson-senet committed:
Fix unit tests

Code review requested changes

snelson-senet committed:
Class B feature branch rebase conflict  fixes

Picked the wrong diff a few times while resolving rebase conflicts.
